src/data/fintype/card_embedding.lean

```diff
@@ -0,0 +1,217 @@
+/-
+Copyright (c) 2021 Eric Rodriguez. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Rodriguez
+-/
+import data.fintype.card
+import data.nat.factorial
+
+/-!
+# Birthday Problem
+
+This file establishes the cardinality of `α ↪ β` in full generality.
+-/
+
+open finset
+
+local notation `|` x `|` := finset.card x
+local notation `‖` x `‖` := fintype.card x
+
+namespace fintype
+
+private def equiv_inj_subtype (α β : Sort*) : {f : α → β // function.injective f} ≃ (α ↪ β) :=
+{ to_fun := λ f, ⟨f.val, f.property⟩,
+  inv_fun := λ f, ⟨f, f.injective⟩,
+  left_inv := λ f, by simp,
+  right_inv := λ f, by {ext, simp} }
+
+instance embedding {α β} [fintype α] [fintype β] [decidable_eq α] [decidable_eq β] :
+  fintype (α ↪ β) :=
+  fintype.of_equiv {f : α → β // function.injective f} (equiv_inj_subtype α β)
```

Comment:
I'd probably put these in `data/fintype/basic`. I really hope there's a nicer way to do this though - since ideally we'd also have `fintype (α →+ β)`, `fintype (α →* β)`, `fintype (α →+* β)` instances, and making a trivial equiv to a subtype each time is tedious.

Reply:
yeah, and this approach also doesn't work for `infinite ↪ finite`, which is obviously empty. i'll have a think about this and see if I can create some nicer approach

Reply:
I don't think you can get around needing an entirely separate instance to handle `finite ↪ finite` and `infinite ↪ finite`.

Reply:
yeah, that's what I was just going to say after messing around with it. `[decidable (infinite α)]` gets around it but that is possibly one of the most useless typeclass arguments in existence

i've also not figured any way to avoid just doing the subtype kerfuffle, i'll work on a separate PR for all the sorts of arrows later and see if they lend themselves to a nicer approach. i'll try cover the list that you covered on your PR

Reply:
okay, so I can write it as this:

`instance embedding' {α β : Type*} [fintype β] [decidable_eq β] [decidable (infinite α)] :
  fintype (α ↪ β) := if h : infinite α then sorry else sorry`

or I can just do two separate instances. my thinking for why this may be okay is that it works well with `open_locale classical`, but I think it'd also need me to create two instances which basically turn `infinite` or `fintype` into `decidable (infinite α)`. or I could just make an `infinite α` instance; how well-used is `infinite` in mathlib?

Reply:
`[decidable (infinite α)]` sounds like a bad idea to me.

Reply:
Just realised - we could use `fintype.of_injective` for the other homs (coe to fun), but that's noncomputable. Thoughts?

Reply:
@eric-wieser, was gonna work on this today; thoughts on `fintype.of_injective`? not sure how much of a big deal noncomputablility is for `→+, →*, →+*`

Reply:
I think a better approach would be to construct the tedious subtype equivs to make it computable. I think noncomputable fintype instances are probably ok in the classical locale, but I wouldn't want to end up using them by accident.